src/pip/_internal/cli/cmdoptions.py

```diff
@@ -884,6 +884,13 @@ def _handle_config_settings(
     help="Don't periodically check PyPI to determine whether a new version "
     "of pip is available for download. Implied with --no-index.",
 )
+default_key_ring_user: Callable[..., Option] = partial(
```

Comment:
Maybe this argument could be usefull for case where companies use internal pypi. And for their users they just could add an additional section for pip.conf:

```text
[global]
index-url = https://pypi-mirror.mangin.com/simple
extra-index-url = https://pypi-private.mangin.com/simple
default-key-ring-user = aleksandr.mangin
```

Reply:
Can this not work if the user name is added to the index URL?

```ini
index-url = https://aleksandr.mangin@pypi-private.mangin.com/simple
```

(I know it does not work now but since we’re doing an overhaul I think we can make it work?)

Reply:
Thank you for your reply.

yep, i checked these scenarious:
1. login in the index-url
2. login in the config
3. login in args.

Would we like to keep `default_key_ring_user`? I think it increase flexiability. And be honest for CI it's the best options. But if you disagree I could revert this change.

P.S. I just would like to know about oppinion of PyPi contibuters how we would like to fix this issue.

Reply:
P.S. When we'll agree on what should be done. I'll add all unit tests ;) I just don't want to do useless job.

Reply:
Personally I feel the extra config is slightly unintuitive. Yes it’s convenient if you know it’s there, but I suspect most people won’t know about it and would be confused seeing the default user kicking in unexpectedly. An explicit username in each URL is more obvious IMO.

Reply:
Okay, let's assume that we remove it.
And now we have two scenarios:
1. Local user tries to install/remove a package
2. Jenkins or any other CI integration tries to install/remove a package.


With first scenario everything is obvious:
1. user can just create a file ~/.pip/pip.conf
2. and add there something like this:
```
[global]
index-url = https://aleksandr.mangin@pypi-mirror.mangin.com/simple
extra-index-url = https://aleksandr.mangin@pypi-private.mangin.com/simple
```
3. and add credentials to keyring.

With jenkins user the situation could be a little bit more difficult.
1. Let's assume that we have a jenkins job that allows us to build some services. So we have many PyPi users.
2. In this case we should receive from the world credentials that we are going to use with PyPi user
3. Generate pip.conf file
```
[global]
index-url = https://system-userN@pypi-mirror.mangin.com/simple
extra-index-url = https://system-userN@pypi-private.mangin.com/simple
```
4. Add credentials to keyring

I think it's a pretty hard to use keyring for system users. But with local users I agree we don't need this parameter.

Action: rollback changes with one more extra-parameter.

What do you think about adding one more source of credentials: Environment variables
1. Let's assume that we have a jenkins job that allows us to build some services. So we have many PyPi users.
2. In this case we should receive from the world credentials that we are going to use with PyPi user
3. Set environment variable PIP_USER/PIP_PASSWORD

Reply:
If you’re going to use environment variables, you can just set the entire `PIP_INDEX_URL` value.

Reply:
Adding a bunch of configuration knobs to pip feels off to me, given that keyring already has support for discovering the username. Is https://github.com/pypa/pip/pull/12748 sufficient?